sdk/python/feast/staging/storage_client.py

```diff
@@ -0,0 +1,312 @@
+#
+# Copyright 2020 The Feast Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import re
+from abc import ABC, ABCMeta, abstractmethod
+from enum import Enum
+from tempfile import TemporaryFile
+from typing import List
+from urllib.parse import ParseResult, urlparse
+
+
+class PROTOCOL(Enum):
+    """
+    Currently supported protocols enum class.
+    """
+
+    GS = "gs"
+    S3 = "s3"
+    LOCAL_FILE = "file"
+
+
+class StorageClient:
```

Comment:
This class seems to be redundant. Its whole purpose is to figure out which `PROTOCOL` maps to which `StagingProtocol` implementation. Maybe we could expose `_protocol_dict` without a wrapping class?

Reply:
@mrzzy Can be done, But one thing that `StorageClient` is doing, adding a place holder and lazy initialization of StagingProtocol. Lazy initialization can be done with just the method `_get_staging_protocol` but this method will have to be invoked before use every invocation. StorageClient cannot be a dependency in class. Let me know what you think. I am open to removing this class and exposing  _protocol_dict

Reply:
> , adding a place holder 

Not sure what you mean here.

> lazy initialization of StagingProtocol. 

Lazy initialization makes sense as the python package already has a terrible start up time and we won't want to worsen it. However I don't think we need a wrapper class to do it, we can just delay calling `__init__()` on the staging protocol classes:
```
storage_protocols = {
    "gcs": GCSProtocol,
    "s3": S3Protocol,
    ...
}
# when we actually use it
staging_protocol = staging_protocol["s3"]()
```

We can wrap this in a `get_staging_protocol(protocol)` to get human readable error messages when a bad protocol is  used while making sure we keep boilerplate to a minimum.

Reply:
Sounds good.